source/adapters/level_zero/device.cpp

```diff
@@ -69,8 +70,49 @@ UR_APIEXPORT ur_result_t UR_APICALL urDeviceGet(
       urPrint("Unknown device type");
       break;
     }
-    if (Matched)
+
+    // We need to filter out composite devices depending on
+    // ZE_FLAT_DEVICE_HIERARCHY value:
+    //   - If COMPOSITE, L0 returns cards as devices. Thus, zeGetRootDevice must
+    //     return nullptr, because they don't have any device higher up in the
+    //     hierarchy.
+    //   - If FLAT,  according to L0 spec, zeGetRootDevice always returns
+    //     nullptr in this mode.
+    //   - If COMBINED, L0 returns tiles as devices, and zeGetRootdevice returns
+    //     the card containing a given tile.
+    //
+    // NOTE: We cannot directly filter out the composite devices here because we
+    // might have the composite device appear earlier than we know we are in
+    // combined mode, so we simply try and infer here if we are in combined
+    // mode, and then, if so, remove composite devices from MatchedDevices.
+    if (!isCombinedMode) {
+      ze_device_handle_t RootDev = nullptr;
+      // Query Root Device
+      ZE2UR_CALL(zeDeviceGetRootDevice, (D->ZeDevice, &RootDev));
```

Comment:
Is this a new entry point? I'm wondering if the GitHub Actions failures could be due to an older driver being installed or something like that?

Reply:
@kbenzie It is a new entry point, but as far as I see, the CI uses level zero v1.15.1 which includes this entry. What I'm seeing is that `zeDeviceGetRootDevice` returns `ZE_RESULT_ERROR_UNSUPPORTED_FEATURE` under certain circumstances (https://github.com/oneapi-src/level-zero/blob/1685d01497428ca4d8b99200972b64685424d5c9/source/lib/ze_libapi.cpp#L463). The CI is running into this, and that makes `ZE2UR_CALL` fail in `UR_CHECK`. I'm not sure how to deal with this, honestly. As of now, I think only PVC should care about this, but I don't think we should have code specific to PVC in UR. Otherwise, we can use `ZE_CALL_NOCHECK` rather than `ZE2UR_CALL` and check manually that the error code is either success or unsupported feature. What do you think? Any other suggestion?

Reply:
Btw, I just checked locally using `ZE_CALL_NOCHECK` and all the tests pass.

Reply:
Done in https://github.com/oneapi-src/unified-runtime/pull/1192/commits/15d2953dd89d60c615b8ab596c6e2761eba83aec